controllers/provisioning_controller.go

```diff
@@ -120,6 +120,63 @@ func (r *ProvisioningReconciler) readProvisioningCR(req ctrl.Request) (*metal3io
 	return instance, nil
 }
 
+//Ensure Finalizer is present on the Provisioning CR when not deleted and
+//delete resources and remove Finalizer when it is
+func (r *ProvisioningReconciler) checkForCRDeletion(baremetalConfig *metal3iov1alpha1.Provisioning) error {
+	// examine DeletionTimestamp to determine if object is under deletion
+	if baremetalConfig.ObjectMeta.DeletionTimestamp.IsZero() {
+		// The object is not being deleted, so if it does not have our finalizer,
+		// then lets add the finalizer and update the object. This is equivalent
+		// registering our finalizer.
+		if !containsString(baremetalConfig.ObjectMeta.Finalizers,
+			metal3iov1alpha1.ProvisioningFinalizer) {
+			// Add finalizer becasue it doesn't already exist
+			baremetalConfig.ObjectMeta.Finalizers = append(baremetalConfig.ObjectMeta.Finalizers,
+				metal3iov1alpha1.ProvisioningFinalizer)
+			if err := r.Client.Update(context.Background(), baremetalConfig); err != nil {
+				return errors.Wrap(err, "failed to update Provisioning CR with its finalizer")
+			}
+		}
+	} else {
+		// The Provisioning object is being deleted
+		if containsString(baremetalConfig.ObjectMeta.Finalizers, metal3iov1alpha1.ProvisioningFinalizer) {
+			err := r.deleteMetal3Resources()
+			if err != nil {
+				return errors.Wrap(err, "failed to delete metal3 pod")
+			}
+			// Remove our finalizer from the list and update it.
+			baremetalConfig.ObjectMeta.Finalizers = removeString(baremetalConfig.ObjectMeta.Finalizers,
+				metal3iov1alpha1.ProvisioningFinalizer)
+			if err = r.Client.Update(context.Background(), baremetalConfig); err != nil {
+				return errors.Wrap(err, "failed to remove finalizer from Provisioning CR")
+			}
+		}
+	}
+	return nil
+}
+
+//Delete Secrets and the Metal3 Deployment objects
+func (r *ProvisioningReconciler) deleteMetal3Resources() error {
+	err := provisioning.DeleteMariadbPasswordSecret(r.KubeClient.CoreV1(), ComponentNamespace)
+	if err != nil {
+		r.Log.V(1).Info("could not delete mariadb password Secret")
+	}
```

Comment:
Need to return inside the if{} block here, otherwise `err` will simply get overwritten.

Reply:
yeah, i was thinking "you are just trying to delete as much as possible and not stress about errors". except you are already taking notFound into account. what other reasons could we have? also if we do anywhere return an error and reenter the Reconcile, how are we going to detect that we are in the middle of a delete? is it just going to say "there is no CR"

Reply:
If we encounter an error while deleteing a resource we return without removing the finalizer. The fact that the finalizer still exists indicates that the cleanup operation is not complete and the Provisioning CR would not be deleted.

Reply:
> Need to return inside the if{} block here, otherwise `err` will simply get overwritten.

We are not returning an error when encountering an issue with deleting a secret. We would just log the fact that we could not delete a secret. We will return an error in the case of other resources and try to delete it next time.